### PR TITLE
Allow defining custom instance types

### PR DESCRIPTION
**Pull request by ichiro999**
We have some customized instance types,  so want to  merge the instance type of local file  and website, rather than directly overwrite

### CODE REVIEW

**Comment by stgraber** [comment]
Hmm, I think I'd prefer allowing a separate file and merging the two together rather than always merging what comes from the website. The reason for that being that we should be able to remove instance type names should a cloud deprecate/remove them.

**Comment by ichiro999** [comment]
> 
thank，your mind is to merge the content of instance_type.yaml with the content of your site.

File: lxd/instance_instance_types.go [hunk 1]
```diff
@@ -110,7 +125,7 @@ func instanceRefreshTypesTask(d *Daemon) (task.Func, task.Schedule) {
 	return f, task.Daily()
 }
 
-func instanceRefreshTypes(ctx context.Context, d *Daemon) error {
+func instanceLoadFromWebsite(ctx context.Context, d *Daemon) map[string]map[string]*instanceType {
```
Comment (tomponline) [inline lxd/instance_instance_types.go:128]: This function should still return an error.

Comment (tomponline) [reply lxd/instance_instance_types.go:128]: And a nil map if an error occurs.

**Comment by tomponline** [comment]
> 

I think @stgraber means that the custom instance types should not be stored in the `shared.CachePath("instance_types.yaml")` file as this is supposed to just be a cache of the instance types from the web site.

Any custom types should be read from a separate file and then merged in memory when loading.

**Comment by ichiro999** [comment]
> > 
> 
> I think @stgraber means that the custom instance types should not be stored in the `shared.CachePath("instance_types.yaml")` file as this is supposed to just be a cache of the instance types from the web site.
> 
> Any custom types should be read from a separate file and then merged in memory when loading.

**Comment by ichiro999** [comment]
> comments

@tomponline thx, it mean that we should provide a directory for customized intsnace type, and "instance_types.yaml" and website is used for default type，lxd merge them to member？

File: lxd/instance_instance_types.go [hunk 2]
```diff
@@ -75,13 +96,14 @@ func instanceRefreshTypesTask(d *Daemon) (task.Func, task.Schedule) {
 	_, hasCancellationSupport := any(&http.Request{}).(util.ContextAwareRequest)
 	f := func(ctx context.Context) {
 		opRun := func(op *operations.Operation) error {
+			defaultAddr := "images.linuxcontainers.org"
```
Comment (tomponline) [inline lxd/instance_instance_types.go:99]: Lets make this a constant at the top of the file called `imagesDomain`

File: lxd/instance_instance_types.go [hunk 3]
```diff
@@ -30,40 +32,59 @@ type instanceType struct {
...
+		return newInstanceType, nil
 	}
 
-	data, err := yaml.Marshal(&instanceTypes)
+	entries, err := os.ReadDir(dir)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	sort.Slice(entries, func(i, j int) bool {
```
Comment (tomponline) [inline lxd/instance_instance_types.go:45]: Newline between each stanza please

File: lxd/instance_instance_types.go [hunk 4]
```diff
@@ -30,40 +32,59 @@ type instanceType struct {
...
-	}
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
 
-	return nil
+		types, err := instanceLoadFromFile(filepath.Join(dir, entry.Name()))
+		if err != nil {
+			logger.Error("read instance types from file error", logger.Ctx{"err": err, "file": entry.Name()})
```
Comment (tomponline) [inline lxd/instance_instance_types.go:56]: Log messages should start with upper case letter

File: lxd/instance_instance_types.go (hunk 3 above)
Comment (tomponline) [reply lxd/instance_instance_types.go:45]: Why are you sorting entries? This should have a comment if its important, so the knowledge is retained. Thanks

File: shared/util.go [hunk 5]
```diff
@@ -233,12 +233,12 @@ func VarPath(path ...string) string {
 // set, this path is $LXD_DIR/cache, otherwise it is /var/cache/lxd.
 func CachePath(path ...string) string {
 	varDir := os.Getenv("LXD_DIR")
-	logDir := "/var/cache/lxd"
```
[the remark this answers is not in the document]
Comment (tomponline) [reply shared/util.go:236]: Good catch though! :)

File: lxd/instance_instance_types.go [hunk 6]
```diff
@@ -228,7 +258,6 @@ func instanceParseType(value string) (map[string]string, error) {
 			}
 		}
 	}
-
```
Comment (tomponline) [inline lxd/instance_instance_types.go:231]: Leave this please

**Review by tomponline — CHANGES_REQUESTED** [review]
I have concerns around the file based approach we are thinking of taking here as it doesn't work like pretty much all other config (which is stored in a database) and thus introduces problems with config distribution in a clustered environment.

Additionally not all of your commits are signed off (which is required, see https://github.com/lxc/lxd/blob/master/CONTRIBUTING.md#developer-certificate-of-origin) and the PR needs to be rebased and structured into logical commits, see https://github.com/lxc/lxd/blob/master/CONTRIBUTING.md#commit-structure

Thanks

**Comment by stgraber** [comment]
Going to take over this one quickly to get it into a usable state for 5.12

**Comment by stgraber** [comment]
Actually, I can't do that as this is not signed-off at all, so it's not just a commit missing it but both commits have it missing.
Going to be closing this for inactivity then as this hasn't been touched since December.

**Pull request closed without merging**